### PR TITLE
server: Add p4-fusion to image

### PR DESCRIPTION
**Pull request by ryanslade**
Copied the steps from our `gitserver` [Dockerfile](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/gitserver/Dockerfile) and tested that I can build
the image locally.

### CODE REVIEW

File: cmd/server/Dockerfile [hunk 1]
```diff
@@ -102,6 +103,39 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
...
+RUN ./generate_cache.sh Release
+RUN ./build.sh
+
+WORKDIR /usr/local/bin
+# Move exe file to /usr/local/bin where other executables are located; delete src directory and archive
+RUN mv /usr/local/bin/p4-fusion-src/build/p4-fusion/p4-fusion /usr/local/bin && \
+    rm -rf /usr/local/bin/p4-fusion-src && \
+    rm /usr/local/bin/v1.5.tar.gz
+# p4-fusion installation completed here
+
```
Comment (sashaostrikov) [inline cmd/server/Dockerfile:138]: Probably we can do `WORKDIR /` at line 146 just in case, as it is done [here](https://github.com/sourcegraph/sourcegraph/blob/9233e2cd9b96bbbbbfb4be2e72543cb41ad9920c/cmd/gitserver/Dockerfile#L95)

**Review by sashaostrikov — APPROVED** [review]
LGTM!

I've just added a small comment about WORKDIR

File: cmd/server/Dockerfile (hunk 1 above)
Comment (ryanslade) [reply cmd/server/Dockerfile:138]: Good idea 👍

**Comment by keegancsmith** [comment]
@ryanslade FYI this greatly increased the time it took for us to build server. Here is a heatmap of build times for the server image, notice the huge step increase when you landed this

image

https://ui.honeycomb.io/sourcegraph/datasets/buildkite/result/sJC2fCavrMg

Do we have any ideas around not needing to recompile all this stuff for every build? eg sharing build artificats? publishing a build somewhere / etc?

**Comment by ryanslade** [comment]
Originally we tried using multi stage builds but we ran into missing dynamic libraries when copying the binary over from the builder image. We could take another look into that though as in the end it was more important to get it working than getting it working fast due to customer time constraints. 

We have the same issue when building our gitserver images.

File: cmd/server/Dockerfile [hunk 2]
```diff
@@ -102,6 +103,39 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
...
 
+# p4-fusion installation
+# Fetching p4 sources archive
+RUN wget https://github.com/salesforce/p4-fusion/archive/refs/tags/v1.5.tar.gz && \
+    mv v1.5.tar.gz /usr/local/bin && \
+    mkdir -p /usr/local/bin/p4-fusion-src && \
+    tar -C /usr/local/bin/p4-fusion-src -xzvf /usr/local/bin/v1.5.tar.gz --strip 1
+
+# We need a specific version of OpenSSL
+RUN wget https://www.openssl.org/source/openssl-1.0.2t.tar.gz && tar -xzvf openssl-1.0.2t.tar.gz
```
Comment (jhchabran) [inline cmd/server/Dockerfile:114]: Drive-by comment: the tarball is not cleaned up after the install.

**Comment by jhchabran** [comment]
> Do we have any ideas around not needing to recompile all this stuff for every build? eg sharing build artificats? publishing a build somewhere / etc?

@keegancsmith We recently added this https://docs.sourcegraph.com/dev/how-to/cache_ci_artefacts to the CI, which is behind a feature flag enabling the stateless agents (Yes, I realized the doc doesn't mention it, I'll fix that!). 

We plan to roll out the stateless agents next week, so that will be available soon. 

I think though that in the present case, dealing with it through Docker by fixing the shared libraries. 

@ryanslade I took a quick look and I'm a bit surprised: 

```
bash-5.0# ldd /bin/bash
	/lib/ld-musl-x86_64.so.1 (0x4000000000)
	libreadline.so.8 => /usr/lib/libreadline.so.8 (0x400195b000)
	libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x4000000000)
	libncursesw.so.6 => /usr/lib/libncursesw.so.6 (0x40019a1000)
bash-5.0# ldd ./p4
/lib/ld-musl-x86_64.so.1: ./p4: Not a valid dynamic program
```

So I wonder what was the error you've seen with the shared libs.

**Comment by keegancsmith** [comment]
Hah, I am literally hacking on this as we speak so expect this build script to be cleaned up. The thing we are building is `p4-fusion` and it does link against libstdc++ which is not part of the alpine base image. Easy to remedy though. Watch this space, I'll link to my initial cleanup of the build script which should be a base we can build on further.
